### PR TITLE
man: Clarify use of fi_cntr_wait wrt errors

### PR DESCRIPTION
**Pull request by shefty**
There is a race condition that exists between an application
reading an error counter and waiting on the counter to
update.  The man page states that a wait is satisfied if
any error occurs; however, an error could occur between the
time the application last read the counter and when wait is
called.

Fix this race by clarifying that wait will be satisfied if
the error counter has changed since the last time the
application read it.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

### CODE REVIEW

**Comment by shefty** [comment]
@sayantansur @j-xiong @jithinjosepkl @a-ilango @goodell @jsquyres -- this is a change/clarification to the cntr wait semantic, which is currently poorly defined and somewhat racy.

File: man/fi_cntr.3.md [hunk 1]
```diff
@@ -217,9 +217,9 @@ return from this call, the counter will be greater than or equal to
 the input threshold value.
 
 If an operation associated with the counter encounters an error, it
-will increment the error value associated with the counter.  Any
-change in a counter's error value will unblock any thread inside
-fi_cntr_wait.
+will increment the error value associated with the counter.  A counter's
+wait object will be signaled if the counter's current error value
+differs from the last value reported by fi_cntr_readerr.
```
Comment (sayantansur) [inline man/fi_cntr.3.md]: Need to think on this a bit ... can "last value" be confusing in a multi-threaded scenario?

Comment (shefty) [inline man/fi_cntr.3.md]: > Need to think on this a bit ... can "last value" be confusing in a multi-
> threaded scenario?

From the perspective of the provider, I don't think so.  The app will need to deal with synchronization if it uses multiple threads to read the error counter and then wants to wait from one or more threads.

**Comment by shefty** [comment]
@sayantansur @j-xiong @jithinjosepkl @a-ilango @goodell @jsquyres -- updated PR.  The man page now documents that reading the error value will atomically reset it to 0.  This aligns better with CQ error handling.  If we accept this change, the providers will need to be updated.  And this can be considered an ABI change.

**Comment by sayantansur** [comment]
Looks good to me. Too bad this is causing an ABI change, but I think it is worthwhile to fix this race.

**Comment by j-xiong** [comment]
👍

**Comment by jithinjosepkl** [comment]
👍

**Comment by a-ilango** [comment]
:+1:

**Comment by goodell** [comment]
👍

**Comment by sayantansur** [comment]
I don't see the cntr_readerr() mentioned in the commits here. did that get dropped?

I was wondering if we should say that cntr_readerr() should "atomically decrement" the reported error value instead of "setting to 0". The issue is that if we say set to zero, then readerr() has to block until some unknown period to ensure there can be no more failures.

`
err = fi_cntr_readerr(ct);

do_wait:
if (ret = fi_cntr_wait(ct)) {
    if (ret < 0 && ret != FI_ETIMEDOUT) {
       /* error occurred on the ct */
       err += fi_cntr_readerr(ct);
       goto do_wait:
    }
}
`

**Comment by j-xiong** [comment]
The "set to zero" only refers to the state at the moment the error counter is retrieved, like how a "read & set" atomic operation works. The actual counter value when the fi_cntr_readerr() function returns can be different.

**Comment by sayantansur** [comment]
That interpretation sounds good. I'm concerned about people misinterpreting the 'set to 0' to imply that the counter is indeed zero when the call returns (which may not be the case). As long as there is enough clarification on the usage of this, sounds good to me.

**Pull request closed without merging**